### PR TITLE
perf(dashboard): Virtualization POC

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
This PR is a proof of concept for dashboard virtualization.
Due to a performance issues on large dashboards (60+ charts in a single tab/view), this PR introduces virtualization in an attempt to reduce the number of elements (DOM nodes) rendered at once.
In the initial state, we render only the charts in current viewport and charts max 1 view height away from current viewport.
When user scrolls down, charts that cross the 1 view height margin get rendered too.
We remove charts from DOM tree when they are at least 4 view heights away from current viewport.
Those values were chosen arbitrarily/experimentally to provide good middle-ground between memory efficiency and good user experience when scrolling up/down.

To sum it up: if chart is not rendered and it's less than 1 view height away from viewport - render it; if chart is rendered and it's more than 4 view heights away from viewport - remove it from DOM

Rerendering charts do NOT trigger new queries - we reuse data fetched on dashboard mount and only redraw the charts.

As of now, this feature is in proof-of-concept faze and should be considered as experimental. It's hidden behind a feature flag `DASHBOARD_VIRTUALIZATION`

Virtualization is disabled for bot workers (like Celery) to ensure that reports and alerts work correctly

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/15073128/194769147-c8060f15-6ea9-403b-aaf0-8270227b94a1.mov

### TESTING INSTRUCTIONS
0. Enable FF `DASHBOARD_VIRTUALIZATION`
1. Open a dashboard with a lot of charts
2. Scroll up and down - verify that only the charts in current viewport (or close to it - 100% margin) are rendered. The rest of them displays a spinner and loads only when user scroll them into view
4. Verify that charts render when user scrolls, but do not send new queries to the backend

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Required feature flags: `DASHBOARD_VIRTUALIZATION`
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by hushaoqing** [comment]
Love this feature.
One suggestion: from my experience, it would be better if it didn't re-render when scrolling back.

**Review by villebro — APPROVED** [review]
LGTM, super nice feature! I kinda agree with @hushaoqing that leaving the charts rendered if they've already been rendered would be nicer. But since this is behind a FF we can probably iterate on these details in forthcoming PRs if there is a good reason not to make that change now.

**Comment by michael-s-molina** [comment]
We currently have `react-virtualized` and `react-window` in our dependencies that provide virtualization. `react-window` is the lighter-weight alternative. Can we use one of them for this problem?

**Comment by kgabryje** [comment]
Thank you for the feedback!
@hushaoqing @villebro Removing charts after scrolling them out of view was a conscious decision in order to reduce the number of nodes in DOM tree. However, I do agree that seeing the loading spinner all the time might be annoying, especially for the charts that take longer to render. It's very easy to change that behaviour, I'll do it later today. If we get some opposite feedback (charts should be removed), I can introduce another feature flag to make it configurable.

@michael-s-molina I considered using `react-window`, but I don't think it's applicable in this case. `react-window` expects a clear list of elements to render, with their heights and widths. The logic of rendering the dashboard is more complex - here we have `DashboardComponent`s recursively rendering other `DashboardComponent`s.
Moreover, even if we managed to restructure dashboard rendering logic to use `react-window`, that would mean that we not only delay drawing the chart, but also running chart queries - as it's the `src/components/Chart.jsx` component that handles running queries and we'd not render it until it's in view.
I think in this case the "manual" implementation of virtualization with `IntersectionObserver` is a better choice

**Comment by michael-s-molina** [comment]
> @hushaoqing @villebro Removing charts after scrolling them out of view was a conscious decision in order to reduce the number of nodes in DOM tree. However, I do agree that seeing the loading spinner all the time might be annoying, especially for the charts that take longer to render. It's very easy to change that behaviour, I'll do it later today. If we get some opposite feedback (charts should be removed), I can introduce another feature flag to make it configurable.

One reason to render only visible items is memory footprint. Many virtualization libraries use that approach. That being said, may I suggest using pre-defined values for the feature flag instead of creating a new one? Like this:

DASHBOARD_VIRTUALIZATION = NONE | VIEWPORT | PAGINATED
NONE = No virtualization is applied
VIEWPORT = Only elements in the viewport are kept in the DOM
PAGINATED = Elements are added to the DOM when they enter the viewport and they are not removed

**Comment by kgabryje** [comment]
> One reason to render only visible items is memory footprint. Many virtualization libraries use that approach. That being said, may I suggest using pre-defined values for the feature flag instead of creating a new one? Like this:
> 
> DASHBOARD_VIRTUALIZATION = NONE | VIEWPORT | PAGINATED NONE = No virtualization is applied VIEWPORT = Only elements in the viewport are kept in the DOM PAGINATED = Elements are added to the DOM when they enter the viewport and they are not removed

Do we have an established pattern for 3-state feature flags?

**Comment by kgabryje** [comment]
>  I'm getting weird performance issues (i.e. the browser freezing) but i'll try it again in a bit and see if it's just a fluke.

If you're getting perf issues with virtualization, but not without it, that might mean that the intersection observer is too expensive to be put on every chart. If that's the case, we might want to put it on `Row` component instead and propagate the visibility prop down to charts...

**Comment by kgabryje** [comment]
> > Do we have an established pattern for 3-state feature flags?
> 
> I'm not aware of one, and it might be worth mentioning that a lot of companies use feature flag manager platforms that prefer to work (or only work) with booleans. Booleans feel safer.

I refactored it like this: `DASHBOARD_VIRTUALIZATION: Literal["NONE", "VIEWPORT", "PAGINATED"] = "NONE"`. It's available in redux state as `common.conf.DASHBOARD_VIRTUALIZATION`. We use a similar pattern with another config option: `DASHBOARD_AUTO_REFRESH_MODE: Literal["fetch", "force"] = "force"`. However, I'm also worried about potential problems with feature flag managers...

**Comment by kgabryje** [comment]
@rusackas @michael-s-molina Updated the PR:
1. Make `DASHBOARD_VIRTUALIZATION` an enum config instead of a feature flag
2. Move `IntersectionObserver` to `Row`. Now we'll have much fewer observers - 1 per row instead of 1 per chart

**Comment by kgabryje** [comment]
Unfortunately I can't start a testenv with virtualization since now it's a config instead of a feature flag, so if you want to test manually, you need to pull my branch to your local setup

**Comment by stephenLYZ** [comment]
Nice feature. Another case to consider is about color consistency. But this scenario is similar to the hidden tab, where the charts are lazy renderer, and this case will be addressed in another PR. cc @geido

**Review by stephenLYZ — APPROVED** [review]

**Comment by mayurnewase** [comment]
Awesome feature, tested locally.

**Comment by michael-s-molina** [comment]
@kgabryje @rusackas Thinking more carefully about the problem, I think we could dynamically choose the strategy (viewport or paginated) based on the number of charts that need to be rendered instead of using feature flags. This way we would have the best of both strategies.

File: superset-frontend/src/utils/isDashboardVirtualizationEnabled.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export enum DASHBOARD_VIRTUALIZATION_MODE {
+  NONE = 'NONE',
+  VIEWPORT = 'VIEWPORT',
+  PAGINATED = 'PAGINATED',
```
Comment (villebro) [inline superset-frontend/src/utils/isDashboardVirtualizationEnabled.ts:23]: Nit: I'm always nagging about this 😄 But I think we should stick to PascalCase enums in our TS codebase: https://www.typescriptlang.org/docs/handbook/enums.html

**Comment by geido** [comment]
> Nice feature. Another case to consider is about color consistency. But this scenario is similar to the hidden tab, where the charts are lazy renderer, and this case will be addressed in another PR. cc @geido

Good catch @stephenLYZ. I think we can use the same approach we would use for the tabs. We would be catching series as they appear and either apply an existing color if that was seen before or assign a new color from the scheme and persist it.

**Review by geido — APPROVED** [review]
Love it! Just one preference from me, I think the options VIEWPORT and PAGINATED are not super clear at first glance. If we set the charts to load on scroll I think LAZY might be a more appropriate keyword.

**Comment by michael-s-molina** [comment]
> Love it! Just one preference from me, I think the options VIEWPORT and PAGINATED are not super clear at first glance. If we set the charts to load on scroll I think LAZY might be a more appropriate keyword.

The problem is that both are lazy 😜. Anyway, I vote for removing the flag/configuration completely and choosing the strategy based on the number of charts that need to be rendered.

**Comment by justinpark** [comment]
Is this safe to support the screen capture from "Reports" worker?
I personally recommend using `react-intersection-observer` which can cover the fallback case of unsupported `IntersectionObserver` (which possibly issue on screen capture from the celery worker)

File: superset-frontend/src/utils/isDashboardVirtualizationEnabled.ts (hunk 1 above)
Comment (ktmud) [reply superset-frontend/src/utils/isDashboardVirtualizationEnabled.ts:23]: +1 on PascalCase. In the latest graphql-codegen CLI, it'd even force the backend enums to be PascalCase by default.

**Comment by kgabryje** [comment]
Thank you all for feedback and suggestions!
@justinpark I'm not sure if delaying queries is a better solution. Queries are initialized in the order in which the charts are placed on the dashboard, so while we usually are exceeding the browser's simultaneous query limit, charts at the top of the page don't really need to wait for the charts at the bottom.
Great catch with the reports! Do you have any idea how to disable virtualization for celery workers?

@michael-s-molina Agreed, but we need to figure out the threshold between the 2 strategies. I was thinking about this: if the chart hasn't been in view yet, draw it when it's 1 view height away from the viewport (like we do now). If the chart has been drawn and viewed, remove it when it's 5 (arbitrarily chosen number) view heights away from the viewport. That way we won't remove charts in regular dashboards, but we will in big dashboards.

**Comment by michael-s-molina** [comment]
> @michael-s-molina Agreed, but we need to figure out the threshold between the 2 strategies. I was thinking about this: if the chart hasn't been in view yet, draw it when it's 1 view height away from the viewport (like we do now). If the chart has been drawn and viewed, remove it when it's 5 (arbitrarily chosen number) view heights away from the viewport. That way we won't remove charts in regular dashboards, but we will in big dashboards.
> WDYT?

Sounds good. We can determine the arbitrary number by displaying a heavy viewport of charts, measuring the consumed memory, and multiplying it by what we consider acceptable.

**Comment by ktmud** [comment]
IIRC, `react-virtualized` has been discontinued by the author in favor of continued support for `react-window`. We should find ways to make our use case work with either `react-window` or `react-intersection-observer`.

File: superset-frontend/src/dashboard/components/gridComponents/Row.jsx [hunk 2]
```diff
@@ -69,6 +81,37 @@ class Row extends React.PureComponent {
...
+
+    this.containerRef = React.createRef();
+    this.observer = null;
+  }
+
+  componentDidMount() {
+    if (
+      isDashboardVirtualizationEnabled(this.props.dashboardVirtualizationMode)
+    ) {
+      this.observer = new IntersectionObserver(
```
Comment (ktmud) [inline superset-frontend/src/dashboard/components/gridComponents/Row.jsx:93]: Should we [disconnect](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/disconnect) and [unobserve](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/unobserve) on unmount? Just to reduce the  chance of memory leak.

Comment (kgabryje) [reply superset-frontend/src/dashboard/components/gridComponents/Row.jsx]: Done

**Comment by kgabryje** [comment]
@michael-s-molina @justinpark @geido @villebro @hushaoqing @rusackas @ktmud Thank you all for feedback and suggestions... and you patience. I updated the PR - now the charts render when they are 1vh away from the viewport, but they're removed from DOM only if they are more than 4vh away from viewport. See the updated summary for more details

**Review by geido — APPROVED** [review]
Thanks for the enhancements!

**Pull request merged**